Print Functionality in Online POS using allow print before pay check

deepak-mnt opened the pull request:
1. "_Allow print before pay_" functionality in Online POS
2. Test Cases to check the print before pay functionality in Online POS
3. Handles the Sales Invoice update/creation in both cases(print before and after submit)

![jmi_pr_1](https://user-images.githubusercontent.com/33081409/34640020-33f60ba0-f311-11e7-9a0b-671a5add467a.gif)

deepak-mnt committed:
Allow draft mode print in online POS

deepak-mnt committed:
[Mod] point_of_sale.js: Allowed print before pay and saved the invoice before print
[New] test_point_of_sale.js: Test Cases to check the print before pay functionality in Online POS

deepak-mnt committed:
[Mod] Ensured that a single POS Sales Invoice is created even when its printed(saved) in Draft mode

deepak-mnt committed:
Minor changes:Removed extra spaces in py code

deepak-mnt committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

deepak-mnt committed:
[Fix] Resolved issues suggested by Codacy

rohitwaghchaure commented on erpnext/selling/page/point_of_sale/point_of_sale.js:
```diff
@@ -50,11 +50,11 @@ erpnext.pos.PointOfSale = class PointOfSale {
 			() => {
 				frappe.dom.unfreeze();
 			},
+			() => this.fetch_pos_profile_data(),
```
No need to write extra method, use existing method.
Set value in the method set_pos_fields, you can add the field allow_print_before_pay in the sales invoice and use same in the POS.
https://github.com/frappe/erpnext/blob/develop/erpnext/accounts/doctype/sales_invoice/sales_invoice.py#L293

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

saurabh-bhosale committed:
[Mod] Refactored the code for fetching allow_print_before_pay check value

saurabh-bhosale committed:
Codacy changes

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

tundebabzy commented on erpnext/selling/page/point_of_sale/point_of_sale.js:
```diff
@@ -50,8 +50,8 @@ erpnext.pos.PointOfSale = class PointOfSale {
 			() => this.setup_company(),
-
 			() => this.make_new_invoice(),
+			() => this.set_form_action(),
```
Let's hide the print button until there is actually an item in the cart.
![peek 2018-01-09 20-30](https://user-images.githubusercontent.com/818803/34740047-70ed4cca-f57e-11e7-8fe6-8fbebbf40abb.gif)

saurabh-bhosale replied:
@tundebabzy Thanks for the input. Have made the changes according to the suggestions in new commits.

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

saurabh-bhosale committed:
[Mod] Allows print only when the check value exists and the cart isn't empty

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

tundebabzy commented on erpnext/selling/page/point_of_sale/point_of_sale.py:
```diff
@@ -88,11 +88,15 @@ def get_conditions(item_code, serial_no, batch_no, barcode):
 
-	doc = frappe.new_doc('Sales Invoice')
+	if islocal_dummy:
+		doc = frappe.get_doc('Sales Invoice',args["name"])
```
Looks like this is causing problems with saving
![peek 2018-01-12 12-54](https://user-images.githubusercontent.com/818803/34879646-3548f5ac-f7ae-11e7-8075-5da829fc715a.gif)

saurabh-bhosale replied:
@tundebabzy  Fixed the bug and refactored the code a bit.

![jmi_pr_1_rev](https://user-images.githubusercontent.com/25682214/34893839-f8d8408a-f804-11e7-80ec-4073576935f7.gif)

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos
Resolved Conflicts:erpnext/accounts/doctype/sales_invoice/sales_invoice.json

saurabh-bhosale committed:
Multiple changes:
[Mod] Used async-await for asynchronous save in draft mode
[Fix] Fixed bug that din't allow to submit a new POS Invoice

tundebabzy commented on erpnext/selling/page/point_of_sale/point_of_sale.js:
```diff
@@ -499,19 +504,26 @@ erpnext.pos.PointOfSale = class PointOfSale {
+		if(this.frm.doc.docstatus == 1 || (this.frm.doc.allow_print_before_pay == 1&&this.frm.doc.items.length>0)){
+			this.page.set_secondary_action(__("Print"), async() => {
+				if(this.frm.doc.docstatus != 1 ){
+					await this.frm.save();
```
I don't know if async/await is ready for mainstream yet

tundebabzy replied:
Ok. I just checked caniuse. I guess this is ok as long as we are not concerned about IE

tundebabzy reviewed (commented):
Works as expected. @netchampfaris is IE a supported browser?

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

saurabh-bhosale committed:
[Fix] Resolved the type issue of 'is_saved' flag variable

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

rohitwaghchaure commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.json:
```diff
@@ -297,6 +297,36 @@
+   "search_index": 0, 
+   "set_only_once": 0, 
+   "unique": 0
+  }, 
```
Why modified date has not changed even after you have added new field, can you rebase?

saurabh-bhosale committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into draft_mode_print_online_pos

rohitwaghchaure commented:
@deepak-mnt 
Closing the PR, kindly make changes and reopen it
